fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/LandPathNodeTypesRegistry.java

```diff
@@ -0,0 +1,277 @@
+
+		// Registers the provider.
+		PathNodeTypeProvider old = NODE_TYPES.put(block, provider);
+
+		if (old != null) {
+			LOGGER.debug("Replaced PathNodeType provider for the block {}", block);
+		}
+	}
+
+	/**
+	 * Registers a {@link DynamicPathNodeTypeProvider} for the specified block, overriding the default block behavior.
+	 *
+	 * <p>A dynamic provider provides the node type basing on the block state, world and position.
+	 * This is more difficult to handle, must be used only if you want to change the node type basing on the position
+	 * of the block in the world, and may degrade the game performances because cannot be optimized but must be
+	 * recalculated at every tick for every entity.
+	 *
+	 * @param block    Block to register.
+	 * @param provider {@link DynamicPathNodeTypeProvider} to associate with the block.
+	 */
+	public static void registerDynamic(Block block, DynamicPathNodeTypeProvider provider) {
+		Objects.requireNonNull(block, "Block cannot be null!");
+		Objects.requireNonNull(provider, "DynamicPathNodeTypeProvider cannot be null!");
+
+		// Registers the provider.
+		PathNodeTypeProvider old = NODE_TYPES.put(block, provider);
+
+		if (old != null) {
+			LOGGER.debug("Replaced PathNodeType provider for the block {}", block);
+		}
+	}
+
+	/**
+	 * Gets the {@link PathNodeType} from the provider registered for the specified block state at the specified position.
+	 *
+	 * <p>If no valid {@link PathNodeType} provider is registered for the block, it returns {@code null}.
+	 * You cannot use this method to retrieve vanilla block node types.
+	 *
+	 * @param state    Current block state.
+	 * @param world    Current world.
+	 * @param pos      Current position.
+	 * @param neighbor Specifies if the block is not a directly targeted block, but a neighbor block in the path.
+	 * @return the custom {@link PathNodeType} from the provider registered for the specified block,
+	 * passing the block state, the world, and the position to the provider, or {@code null} if no valid
+	 * provider is registered for the block.
+	 */
+	@Nullable
+	public static PathNodeType getPathNodeType(BlockState state, BlockView world, BlockPos pos, boolean neighbor) {
+		Objects.requireNonNull(state, "BlockState cannot be null!");
+		Objects.requireNonNull(world, "BlockView cannot be null!");
+		Objects.requireNonNull(pos, "BlockPos cannot be null!");
+
+		// Gets the node type provider for the block.
+		PathNodeTypeProvider provider = NODE_TYPES.get(state.getBlock());
+
+		//Gets the node type from the registered provider.
+		if (provider instanceof DynamicPathNodeTypeProvider dynamicProvider) {
+			//This can return null too.
+			return dynamicProvider.getPathNodeType(state, world, pos, neighbor);
+		} else if (provider instanceof StaticPathNodeTypeProvider staticProvider) {
+			//This can return null too.
+			return staticProvider.getPathNodeType(state, neighbor);
+		}
+
+		//If no valid provider is found returns null.
+		return null;
+	}
+
```

Comment:
I think we can have a lower-level API here, just because I don't like all the boilerplate that we need:

(sketch)
```java
/**
 * Return the implementation of the provider registered for a specific block, or {@code null} if nothing is registered.
 *
 * <p>If non-null, the returned object can either be a {@link StaticPathNodeTypeProvider} or a {@link DynamicPathNodeTypeProvider} at the moment depending on what was registered.
 * More kinds of providers might be added if the API is expanded in the future, so make sure not to fail without exceptions if another type of object is returned.
 */
@Nullable
public static Object getProvider(Block block) {
    Objects.requireNonNull(block);
    return NODE_TYPES.get(block);
}
```

Reply:
Ok done.